Source/WebKit/Platform/IPC/ArgumentCoders.h

```diff
@@ -519,48 +518,29 @@ struct TupleEncoder {
         if constexpr (sizeof...(Indices) > 0)
             (encoder << ... << std::get<Indices>(tuple));
     }
-};
-
-template<typename... Elements> struct TupleDecoder;
-
-template<typename Type, typename... Types>
-struct TupleDecoder<Type, Types...> {
-    template<typename Decoder>
-    static std::optional<std::tuple<Type, Types...>> decode(Decoder& decoder)
-    {
-        std::optional<Type> optional;
-        decoder >> optional;
-        if (!optional)
-            return std::nullopt;
-
-        std::optional<std::tuple<Types...>> remainder = TupleDecoder<Types...>::decode(decoder);
-        if (!remainder)
-            return std::nullopt;
-
-        return std::tuple_cat(std::make_tuple(WTFMove(*optional)), WTFMove(*remainder));
-    }
-};
 
-template<>
-struct TupleDecoder<> {
     template<typename Decoder>
-    static std::optional<std::tuple<>> decode(Decoder&)
+    static std::optional<std::tuple<Elements...>> decode(Decoder& decoder)
     {
-        return std::make_tuple();
+        return decode(decoder, std::tuple<> { }, std::index_sequence_for<> { });
     }
-};
 
-template<typename... Elements> struct ArgumentCoder<std::tuple<Elements...>> {
-    template<typename Encoder>
-    static void encode(Encoder& encoder, const std::tuple<Elements...>& tuple)
+    template<typename Decoder, typename OptionalTuple, size_t... Indices>
+    static std::optional<std::tuple<Elements...>> decode(Decoder& decoder, OptionalTuple&& optionalTuple, std::index_sequence<Indices...>)
     {
-        TupleEncoder<Elements...>::encode(encoder, tuple);
-    }
+        constexpr size_t Index = sizeof...(Indices);
+        static_assert(Index == std::tuple_size_v<OptionalTuple>);
 
-    template<typename Decoder>
-    static std::optional<std::tuple<Elements...>> decode(Decoder& decoder)
-    {
-        return TupleDecoder<Elements...>::decode(decoder);
+        if constexpr (Index < sizeof...(Elements)) {
+            std::optional<std::tuple_element_t<Index, std::tuple<Elements...>>> optional;
```

Comment:
I did try invoking `decoder.template decode<>()` but it got stuck on `Daemon::Decoder` using this ArgumentCoder specialization and not providing the method. But yes, it would be more efficient to use.

I'll land this and then enhance `Daemon::Decoder`.

I am building on Mac, I'll look into adding the unit tests too.